File: Algorithms/src/2116. Check if a Parentheses String Can Be Valid/Code/solution.py
class Solution:
    def canBeValid(self, s: str, locked: str) -> bool:
        if len(s) % 2 != 0:
            return False  # Odd length can't be balanced
        
        open, flexible = 0, 0
        # Left-to-right pass
        for i in range(len(s)):
            if locked[i] == '1':
                open += 1 if s[i] == '(' else -1
            else:
                flexible += 1
            if open + flexible < 0:
                return False
        
        open, flexible = 0, 0
        # Right-to-left pass
        for i in range(len(s) - 1, -1, -1):
            if locked[i] == '1':
                open += 1 if s[i] == ')' else -1
            else:
                flexible += 1
            if open + flexible < 0:
                return False
        
        return True

File: Algorithms/src/2116. Check if a Parentheses String Can Be Valid/Code/solution.cpp
class Solution
{
public:
    bool canBeValid(string s, string locked)
    {
        if (s.size() % 2 != 0)
            return false; // Odd length can't be balanced

        int open = 0, flexible = 0;
        // Left-to-right pass
        for (int i = 0; i < s.size(); i++)
        {
            if (locked[i] == '1')
            {
                open += (s[i] == '(' ? 1 : -1);
            }
            else
            {
                flexible++;
            }
            if (open + flexible < 0)
                return false;
        }

        open = 0, flexible = 0;
        // Right-to-left pass
        for (int i = s.size() - 1; i >= 0; i--)
        {
            if (locked[i] == '1')
            {
                open += (s[i] == ')' ? 1 : -1);
            }
            else
            {
                flexible++;
            }
            if (open + flexible < 0)
                return false;
        }

        return true;
    }
};
